sound/soc/sof/intel/hda-dai.c

```diff
@@ -330,7 +330,8 @@ static int hda_link_pcm_trigger(struct snd_pcm_substream *substream,
 		if (ret < 0)
 			return ret;
 		stream_tag = hdac_stream(link_dev)->stream_tag;
-		snd_hdac_ext_link_clear_stream_id(link, stream_tag);
+		if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+			snd_hdac_ext_link_clear_stream_id(link, stream_tag);
```

Comment:
@RanderWang But why? We need link and host DMA for capture as well, why the different treatment?

Reply:
> ... I think this is the main issue. What I still don't fully understand is
> that why hda_link_stream_assign() doesn't find the right stream_tag
> with current code.

@RanderWang I've been trying to reproduce this by on WHL but to no avail. 
While I still don't fully get how stream_assign() is supposed to work,
it would seem your patch is at least the safer option.

It's confusing that stream_assign() basicly returns the first available hdac-ext stream, and if there
are any non-opened streams in the bus (with matching direction), the caller will get a random stream back. The assumption made is that this only happens in non-host-dma cases, but it's very hard to verify from code review this assumption is ok.

Reply:
@kv2019i  very good finding ! 
the author of this function is @ranj063 ,  but I can share my understanding.
In ASoC, the calling sequence of playback is :   FE stream open --- BE stream hw_param --- FE stream hw_param.  The hdac_ext stream for FE stream is allocated in hda_dsp_pcm_open firstly, and for BE stream is done in hda_link_hw_params later. So in stream_assign, the assumption is if there are FE & BE for a pcm stream, they will be combined when searching list from head to tail.

This is not reasonable, and the allocation is not optimized, .e.g 
(1) pcm1 FE open + BE hw_param// both stream tag 2 is allocated for the FE & BE
(2) pcm2 FE open// stream tag 3 is allocated for the FE
(3) pcm1 FE & BE release //both stream tag 2 is released for the FE & BE
(4) pcm2 BE hw_param // stream tag 2 is allocated for this BE for the FE with stream tag 2 is not used now. But the optimized way is allocating stream tag 3 to the BE.

We should check all the streams to find whether there is a stream used by the FE of this BE, if there is none, the BE should be a hostless stream.

My patch is not for this case. I think we should not release stream for BE when doing S3.  the hda stream for FE in SOF driver is still active and the stream tag is still used by hda codec driver, so we should let ASoC to release all the streams in normal way

Reply:
I don't know whether you know the background of hda dma.  There are two hda dma engines : host
dma and link dma. For SOF,  dsp is enabled and hda dma is set to decoupled mode. host dma channel number is not necessary equal to link dma for a pcm stream. Also there is N host dma linked to M link dma case.  But for a pcm with only one FE and BE, it is best to use the same channel number(stream tag) for a hardware requirement( it is a long story...).